File: src/main/java/com/timeback/ui/feature/apps/AppManagementViewModel.java
package com.timeback.ui.feature.apps;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.timeback.ui.domain.gateway.FeatureGateway;
import com.timeback.ui.domain.model.ActionResult;
import com.timeback.ui.domain.model.AppClassification;
import com.timeback.ui.domain.model.ScreenState;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

/**
 * UI-04 앱 관리 ViewModel.
 * 설치된 앱 목록과 기본 분류를 관리한다.
 */
@HiltViewModel
public class AppManagementViewModel extends ViewModel {

    private final FeatureGateway gateway;
    private final MutableLiveData<ScreenState> _state = new MutableLiveData<>(new ScreenState.Initial());
    public final LiveData<ScreenState> state = _state;

    private final MutableLiveData<ActionResult> _actionResult = new MutableLiveData<>();
    public final LiveData<ActionResult> actionResult = _actionResult;

    @Inject
    public AppManagementViewModel(FeatureGateway gateway) {
        this.gateway = gateway;
        loadApps();
    }

    public void loadApps() {
        _state.setValue(new ScreenState.Loading());
        gateway.readApps().thenAccept(_state::postValue);
    }

    public void changeClassification(String packageName, AppClassification classification) {
        gateway.changeDefaultClassification(packageName, classification).thenAccept(_actionResult::postValue);
    }
}

File: src/main/java/com/timeback/ui/feature/apps/AppManagementViewModel.kt
package com.timeback.ui.feature.apps

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.timeback.ui.domain.gateway.FeatureGateway
import com.timeback.ui.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI-04 앱 관리 ViewModel
 * CT-04 §7: ReadApps, ChangeDefaultClassification
 */
@HiltViewModel
class AppManagementViewModel @Inject constructor(
    private val gateway: FeatureGateway
) : ViewModel() {

    private val _state = MutableStateFlow<ScreenState>(ScreenState.Initial)
    val state: StateFlow<ScreenState> = _state.asStateFlow()

    private val _actionResult = MutableStateFlow<ActionResult?>(null)
    val actionResult: StateFlow<ActionResult?> = _actionResult.asStateFlow()

    init {
        loadApps()
    }

    fun loadApps() {
        viewModelScope.launch {
            _state.value = ScreenState.Loading
            _state.value = gateway.readApps()
        }
    }

    fun changeClassification(packageName: String, classification: AppClassification) {
        viewModelScope.launch {
            val result = gateway.changeDefaultClassification(packageName, classification)
            _actionResult.value = result
            if (result is ActionResult.Success) {
                _state.value = gateway.readApps()
            }
        }
    }

    fun clearActionResult() {
        _actionResult.value = null
    }
}
